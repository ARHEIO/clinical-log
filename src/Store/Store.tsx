import React, { ReactElement, useState, createContext } from 'react';
import { ThemeProvider, createMuiTheme, Theme } from '@material-ui/core/styles';

import createBreakpoints from '@material-ui/core/styles/createBreakpoints';

/**
 * xs, extra-small: 0px
 * sm, small: 600px
 * md, medium: 960px
 * lg, large: 1280px
 * xl, extra-large: 1920px
 */
const breakpoints = createBreakpoints({});

const getPixelSize = (x: number): string => `${x * 8}px`;

const theme: Theme = createMuiTheme({
  palette: {
    primary: {
      main: '#546A7B',
    },
    secondary: {
      main: '#87AEB5',
    },
  },
  overrides: {
    MuiButton: {
      root: {
        fontSize: '14px',
        minWidth: '100%',
        minHeight: getPixelSize(6),
        [breakpoints.up('md')]: {
          minWidth: getPixelSize(25),
        },
      },
    },
  },
});

export const AuthContext = createContext<any>({ isAuthed: false });

const Store = (props: {children: any}): ReactElement => {
  const { children } = props;
  const [authContext, setAuthContext] = useState({ isAuthed: false });

  return (
    <>
      <AuthContext.Provider value={[authContext, setAuthContext]}>
        <ThemeProvider theme={theme}>
          {children}
        </ThemeProvider>
      </AuthContext.Provider>
    </>
  );
};

export default Store;
