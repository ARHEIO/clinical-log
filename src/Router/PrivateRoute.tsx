/* eslint-disable react/jsx-props-no-spreading */
import React, { useContext, ReactElement } from 'react';
import { Route, Redirect } from 'react-router-dom';
import { AuthContext } from '../Store/Store';

// TODO: find a better way to implement private routes than playing the assignment game

const PrivateRoute = (props: any): ReactElement => {
  const { component, ...rest } = props;
  const Component = component;
  const [authContext] = useContext(AuthContext);

  return (
    <Route
      {...rest}
      render={(innerProps): ReactElement => (
        authContext.isAuthed === true
          ? <Component {...innerProps} />
          : <Redirect to="/newsfeed" />
      )}
    />
  );
};

export default PrivateRoute;
