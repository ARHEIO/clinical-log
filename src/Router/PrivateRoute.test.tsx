import React, { ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import pretty from 'pretty';
import { MemoryRouter } from 'react-router-dom';
import PrivateRoute from './PrivateRoute';
import { AuthContext } from '../Store/Store';

const mockComponent = (): ReactElement => (<div>An Authed Component</div>);

jest.mock('react-router-dom', () => {
  const domOriginalModule = jest.requireActual('react-router-dom');
  return {
    __esModule: true,
    ...domOriginalModule,
    /* eslint-disable-next-line react/prop-types, react/display-name */
    Redirect: ({ to }): ReactElement => (
      <div>
        <p>
          This is the redirect:
          {to}
        </p>
      </div>
    ),
  };
});

const renderComponent = (isAuthed = false): RenderResult => (
  render(
    <AuthContext.Provider value={[{ isAuthed }]}>
      <MemoryRouter initialEntries={['/']}>
        <PrivateRoute component={mockComponent} />
      </MemoryRouter>
    </AuthContext.Provider>,
  )
);

describe('PrivateRoute', () => {
  it('should render redirect when auth is false', () => {
    const { container } = renderComponent();
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
  it('should render component when auth is true', () => {
    const { container } = renderComponent(true);
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
