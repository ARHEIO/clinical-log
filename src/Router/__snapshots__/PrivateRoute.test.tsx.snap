// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute should render component when auth is true 1`] = `"<div>An Authed Component</div>"`;

exports[`PrivateRoute should render redirect when auth is false 1`] = `
"<div>
  <p>This is the redirect:/newsfeed</p>
</div>"
`;
